### ISSUE TITLE
Selftests for kernel symbol name length limitation

### ISSUE THREAD

**Issue by fbq** [Rust-for-Linux/linux#504]
Rust support [enlarged](https://lore.kernel.org/lkml/20210704202756.29107-2-ojeda@kernel.org/) the length limitation of kernel symbols. We should provide selftests showing the longest name is supported, a simple one would be like:

```diff
diff --git a/kernel/kallsyms.c b/kernel/kallsyms.c
index 4067564ec59f..e356a10bcb51 100644
--- a/kernel/kallsyms.c
+++ b/kernel/kallsyms.c
@@ -865,8 +865,22 @@ static const struct proc_ops kallsyms_proc_ops = {
        .proc_release   = seq_release_private,
 };

+#define DI(name) n##name##name
+#define DDI(name) DI(n##name##name)
+#define DDDI(name) DDI(n##name##name)
+#define DDDDI(name) DDDI(n##name##name)
+#define DDDDDI(name) DDDDI(n##name##name)
+
+/* Generate a symbol whose name length is 511 */
+#define LONGEST_NAME DDDDDI(g1h2i3j4k5l6m7n)
+
+extern int LONGEST_NAME(void) { return 0; }
+
 static int __init kallsyms_init(void)
 {
+       /* Showing that the longest name is supported */
+       BUG_ON(!kallsyms_lookup_name(__stringify(LONGEST_NAME)));
+
        proc_create("kallsyms", 0444, NULL, &kallsyms_proc_ops);
        return 0;
 }

```
However, we may want something better as [suggested](https://github.com/Rust-for-Linux/linux/pull/503#issuecomment-931279345) by @ojeda . Therefore create this issue to track work on this.

**Comment by sergiocollado** [comment]
@fbq  @ojeda  I will have a go to try to write the selftest for this issue, it it is ok.

**Comment by ojeda** [comment]
Of course!

**Comment by fbq** [comment]
> @fbq @ojeda I will have a go to try to write the selftest for this issue, it it is ok.

FWIW, I had a go myself when I was playing around `rust_selftests.rs`, at that time I used procedure macro to generate the longest symbol. I cannot find the code though ;-(

**Comment by ojeda** [comment]
Now that you mention it Boqun, I recall I had started some code on that too -- I have rebased and uploaded it to https://github.com/ojeda/linux/commit/0329a987c831ac66ade14daf948e10c2971655c7

In my case I went the C route, since I thought testing the longest symbol would be worth it in all cases, not just for Rust, so I went with a hostprog to generate the symbol.

Hope that helps!

**Comment by sergiocollado** [comment]
For testing this with selftests I was thinking to write a test_module, define and export a symbol with the max length, and then check if the symbol is present in the /proc/kallsyms. What do you think about this approach @fbq @ojeda ? Would it be ok?

**Comment by sergiocollado** [comment]
@ojeda @fbq , I have created a draft PR here: #923 . Please have a look, any feedback will be welcomed. 

There are still some open points. For example, the placement of the test could be in other folder. I placed it in that folder (/lib), because there is where the most of the other test modules are.  Or maybe the code could be written in another way. 

I have written the test module, according to the guide: https://docs.kernel.org/dev-tools/kselftest.html#test-module

And the kprobes - kallsyms used in the code, are due that since Linux 5.7.7 kallsyms is not exported (this is summarized here: https://nskernel.gitbook.io/kernel-play-guide/accessing-the-non-exported-in-modules#for-using-after-5.7.7 ). And the 
solution I took is the described here: https://sysprog21.github.io/lkmpg/ , at "10 System Calls". 

Also if you could test it indepedently it would be great. (just to confirm it works)

**Comment by sergiocollado** [comment]
I have created this PR #956, which uses kunit test, to verify the longest symbol name length

**Comment by sergiocollado** [comment]
I think this issue is complete, a Kunit test checking the longest symbol, has already been merged: https://lore.kernel.org/rust-for-linux/CABVgOSm=5Q0fM6neBhxSbOUHBgNzmwf2V22vsYC10YRBT=kN1g@mail.gmail.com/T/#t

**Comment by ojeda** [comment]
Yeah -- if related issues appear, we can open new ones. Thanks Sergio for keep pushing for this!

**Comment by fbq** [comment]
@sergiocollado Thanks for doing the work. Happy to see it gets merged, however, there seems to be a few problem needed to resolved:

https://lore.kernel.org/lkml/66f3f697-530a-4d22-979a-1472c8e8de3a@app.fastmail.com/
https://lore.kernel.org/lkml/20250328112156.2614513-1-arnd@kernel.org/

Could you take a look into this and discuss some potential resolutions? Thanks!

**Comment by sergiocollado** [comment]
@fbq Hello, 

 I have already had a look at it, and there is a fix proposal: make the test dependent on !CONFIG_PREFIX_SYMBOLS . I sent a mail to David G. to coordinate with him the fix. 

 Some issues were already expected before hand (https://lore.kernel.org/rust-for-linux/CABVgOSm=5Q0fM6neBhxSbOUHBgNzmwf2V22vsYC10YRBT=kN1g@mail.gmail.com/). 

Also the message "Please increase KSYM_NAME_LEN both in kernel and kallsyms.c\n" was already present in the kernel...  the test is only making it obvious. 

 I didn't mention anything of this, because this is not related to RfL.

Thanks for the heads up!
